extensions/amp-playbuzz/0.1/amp-playbuzz.js

```diff
@@ -0,0 +1,290 @@
+    iframe.setAttribute('allowfullscreen', 'true');
+    iframe.src = this.generateEmbedSourceUrl_();
+
+    this.listenToPlaybuzzItemMessage_('resize_height',
+      utils.debounce(this.itemHeightChanged_.bind(this), 100));
+
+    this.element.appendChild(this.getOverflowElement_());
+
+    this.applyFillContent(iframe);
+    this.element.appendChild(iframe);
+
+    return this.iframePromise_ = this.loadPromise(iframe).then(() => {
+
+      this.attemptChangeHeight(this.itemHeight_).catch(() => {/* die */});
+
+      const unlisten = this.getViewport().onChanged(
+        utils.debounce(this.sendScrollDataToItem_.bind(this), 250));
+      this.unlisteners_.push(unlisten);
+    });
+  }
+
+  /** @return {!Element} @private */
+  createPlaybuzzLoader_() {
+    const doc = this.element.ownerDocument;
+    const createElement = utils.getElementCreator(doc);
+
+    const loaderImage = createElement('img', 'pb_feed_anim_mask');
+    loaderImage.src = logo;
+
+    const loaderText = createElement('div', 'pb_feed_loading_text');
+    loaderText.textContent = 'Loading...';
+
+    const loadingPlaceholder =
+      createElement('div', 'pb_feed_placeholder_container',
+        createElement('div', 'pb_feed_placeholder_inner',
+          createElement('div', 'pb_feed_placeholder_content', [
+            createElement('div', 'pb_feed_placeholder_preloader', loaderImage),
+            loaderText,
+          ])));
+
+    return loadingPlaceholder;
+  }
+
+  /**
+   * @param {number} height
+   */
+  itemHeightChanged_(height) {
+
+    if (isNaN(height) || height === this.itemHeight_) {
+      return;
+    }
+
+    this.itemHeight_ = height; //Save new height
+
+    this.attemptChangeHeight(this.itemHeight_).catch(() => {/* die */});
+  }
+
+
+  /**
+   * @param {string} messageName
+   * @param {function} handler
+   */
+  listenToPlaybuzzItemMessage_(messageName, handler) {
+    const unlisten = events.listen(this.win, 'message',
+      event => utils.handleMessageByName(this.iframe_,
+        event, messageName, handler));
+    this.unlisteners_.push(unlisten);
+  }
+
+  /**
+   *
+   * Returns the composed embed source url
+   * @returns {string} url
+   *
+   */
+  generateEmbedSourceUrl_() {
+    const itemSrc = parseUrl(this.item_);
+    const winUrl = this.win.location;
+    const params = {
+      itemUrl: removeFragment(itemSrc.href).replace(itemSrc.protocol, ''), //remove scheme (cors) & fragment
+      relativeUrl: itemSrc.pathname,
+      displayItemInfo: this.displayItemInfo_,
+      displayShareBar: this.displayShareBar_,
+      displayComments: this.displayComments_,
+      parentUrl: removeFragment(winUrl.href),
+      parentHost: winUrl.hostname,
+    };
+
+    const embedUrl = utils.composeEmbedUrl(params);
+    return embedUrl;
+  }
+
+  sendScrollDataToItem_(changeEvent) {
+    const viewport = this.getViewport();
+
+    const scrollingData = {
+      event: 'scroll',
+      windowHeight: changeEvent.height,
+      scroll: changeEvent.top,
```

Comment:
Hey Dima, we need the scrolling events to understand where our iframe's content is relative to the top of the page, and we need this information to be up-to-date. So, understanding that sending a postMessage to the iframe on every scroll is too much, we're sending this postMessage using debounce.
Using intersection observer seems like a really cool idea, since we're getting all the data we need in the callback (and therefore don't have to query for the iframe's offset, for example), but in order to receive this event as often as we need it, we will have to provide the observer with a pretty long threshold array. In this scenario, won't it be a little inefficient?